### PR TITLE
Blog App : Add Authorization Rules

### PR DESCRIPTION
**Pull request by ehmaddd**
### This Pull Request has the following features :

- Installed CanCanCan in your project.
- Added a role column to the users table using migration.

--------------------------------------------------------------------------------------------------

**Using CanCanCan for authorization :**

- A user can delete a post if it is theirs or if they have an admin role (column role has value "admin"). 
   - Implemented the post deleting functionality.
   - Added the "Delete" button to the view which only authorized users can see it.
- A user can delete a comment if it is theirs or if they have an admin role (column role has value "admin"). 
   - Implemented the comment deleting functionality.
   - Added the "Delete" button to the view and make sure that only authorized users can see it.

--------------------------------------------------------------------------------------------------

General Requirements :

- There is no linter error.
- Used correct Gitflow.
- Documented work in a professional way.
- Followed the list of best practices for Ruby

### CODE REVIEW

File: app/models/ability.rb [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+    user ||= User.new # Guest user (not logged in)
+
+    can :destroy, Post, author_id: user.id # Allow users to delete their own posts
+    can :destroy, Comment, author_id: user.id # Allow users to delete their own comments
+
+    return unless user.role == 'admin' # Check if the user has an admin role
+
+    can :destroy, Post # Allow admins to delete any post
+    can :destroy, Comment # Allow admins to delete any comment
+  end
```
Comment (VICTOR-CHUKWU) [inline app/models/ability.rb]: - [x] You have done an excellent job, but here you are giving the admin and user the right only to destroy, but no right to read, like and create a comment, also admin should be able to manage all activities. so you can look at this code snippet and revamp your method
```
 if user.admin? 
         can :manage, :all
         else
         can :manage, Post, author: user  # owner of post can perform all view, update or destroy own post if Post.author == user
         can :manage, Comment, author: user
         can :read, :all
         end
```

File: app/controllers/posts_controller.rb [hunk 2]
```diff
@@ -13,6 +13,19 @@ def create
...
+
+    if @post.destroy
+      flash[:notice] = 'Post was successfully deleted.'
+    else
+      flash[:alert] = 'Unable to delete the post.'
+    end
+
+    redirect_to user_path(@post.author)
+  end
+
```
Comment (VICTOR-CHUKWU) [inline app/controllers/posts_controller.rb:29]: - [x] kindly add user authorization at the beginning of this posts_controller class `load_and_authorize_resource`

**Review by VICTOR-CHUKWU — CHANGES_REQUESTED** [review]
Hi @ehmaddd  👋  

Good job so far!
There are some issues that you still need to work on to go to the next project but you are almost there!

almost

## Highlights
- Good PR
- Good commit
- All linters passed


## Required Changes ♻️

_Check the comments under the review._

## Optional suggestions
_Every comment with the [OPTIONAL] prefix is not crucial enough to stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better._

Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561) you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/ability.rb (hunk 1 above)
Comment (ehmaddd) [reply app/models/ability.rb]: Added as instructed

File: app/controllers/posts_controller.rb (hunk 2 above)
Comment (ehmaddd) [reply app/controllers/posts_controller.rb:29]: Done as instructed

File: app/models/ability.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+class Ability
+  include CanCan::Ability
+
+  def initialize(user)
+    user ||= User.new # Guest user (not logged in)
+
+    if user.admin?
```
Comment (danmainah) [inline app/models/ability.rb:7]: - [x] @Team great work implementing the change requested, however, you shifted from using the string `admin` to `admin?` method, which is not defined in your app leading to the error below when I try to access the posts index page, you can solve this by adding the admin? the method in your user model 
```
 def admin?
     role == 'admin' ? true : false
 end
````
image

**Review by danmainah — CHANGES_REQUESTED** [review]
Hi @ehmaddd and @Patfarmurs , 

Great work on making the changes requested by a previous reviewer 👏  👏 👏
You've done well implementing some of the requested changes, but there are still some which aren't addressed yet.


 ## Highlights
✔️  Implemented the change requested to ensure the roles are well defined in the ability class.
✔️   Implemented the change requested to ensure the use of CanCanCan for Post authorization in `Post Controller`.

 # Required Changes ♻️
_Check the comments under the review._

## Optional suggestions

_Every comment with the [OPTIONAL] prefix won't stop the approval of this PR. However, I strongly recommend you to take them into account as they can make your code better. Some of them were simply missed by the previous reviewer and addressing them will really improve your application._


Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

**_Please, do not open a new Pull Request for re-reviews. You should use the same Pull Request submitted for the first review, either valid or invalid unless it is requested otherwise._**


------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

File: app/models/ability.rb (hunk 3 above)
Comment (ehmaddd) [reply app/models/ability.rb:7]: Added admin? method as per instructions

**Review by hiromataba — APPROVED** [review]
# Status: Approved :tada: :tada: :tada: 
Hi @ehmaddd and @Patfarmurs  :wave: :wave: 

200w  

Your project is complete!  There is nothing else to say other than... it's time to merge it :shipit: 
Congratulations! 🎉 



Cheers and Happy coding!👏👏👏

Feel free to leave any questions or comments in the PR thread if something is not 100% clear.
Please, remember to tag me in your question so I can receive the notification.

------
_As described in the [Code reviews limits policy](https://microverse.zendesk.com/hc/en-us/articles/1500004088561)  you have a limited number of reviews per project (check the exact number in your Dashboard). If you think that the code review was not fair, you can request a second opinion using [this form](https://airtable.com/shrQAqnBwek5a0O0s)._

**Pull request merged**